strapi/strapi
# Refactor content manager field sizes management

### What does it do?

- Introduce a `fieldSizes` service in the Content Manager backend. This service uses an object as the source of truth for the rules that manage (whether they can be resized, what their default should be).
- Introduce a `content-manager/init` endpoint in order to make the `fieldSizes` object accessible globally in the content manager frontend
- Refactor the `useModels` hooks into a `useContentManagerInitData` hook
- Instead of making 3 requests to `/init`, `/content-types` and `components`, the new hook only fetches the init data, where the content types and components data are also available

### Why is it needed?

#### fieldSizes object

Before, the rules (default + resize) for each field type were duplicated in several places in the admin and server. This was error-prone. Having an object as the source of truth makes things more declarative and understandable. It will also enable us to let custom field developers override these rules in the future.

#### useModels refactoring

- In our internal RFC, we suggested fetching the `/init` in a separate `useContentManagerInitData` hook that lives besides `useModels`, and storing the data inside the `content-manager_app` object in Redux. However, that object already has a `status` key that managed the 2 other requests. Since we needed the 2+1 requests to be successful before rendering, it was more convenient to tie them all together in the same reducer/hook.
- Once having the `useContentManagerInitData`, I found that merging all requests into init was better for performance, since we avoid API calls, and the logic on the backend could easily be shared via the services. This can be reverted if needed, all related changes are in [this commit](https://github.com/strapi/strapi/pull/16387/commits/f57ddb4cef9b815396828e0169a9cc4eecb45a11).
- It also makes it easier to describe the point of the `/init` endpoint: it's everything that's need to let the content manager start rendering.
- Why not the delete /content-types and /components endpoints? They are used by [_another_ `useModels` hook](https://github.com/strapi/strapi/blob/f57ddb4cef9b815396828e0169a9cc4eecb45a11/packages/core/admin/admin/src/hooks/useModels/index.js#L5)


### How to test it?

- In the "configure the view" page, try resizing fields. It should not be possible for rich text, dynamic zones, component, and json fields. For all the other field types, all sizes should be available
- In the "configure the view", try deleting a field. When you re-add it, it should have the right default size.
- For all your content types where you didn't configure the view, you should see the same layout as you did before (number takes 4/12 columns by default, text 6/12, etc.)
